// @copilotkit/shared/dist/utils/index.mjs
import {
  COPILOT_CLOUD_ERROR_NAMES,
  ConfigurationError,
  CopilotKitAgentDiscoveryError,
  CopilotKitApiDiscoveryError,
  CopilotKitError,
  CopilotKitErrorCode,
  CopilotKitLowLevelError,
  CopilotKitMisuseError,
  CopilotKitRemoteEndpointDiscoveryError,
  CopilotKitVersionMismatchError,
  ERROR_CONFIG,
  ERROR_NAMES,
  MissingPublicApiKeyError,
  ResolvedCopilotKitError,
  Severity,
  UpgradeRequiredError,
  getPossibleVersionMismatch,
  isMacOS,
  parseJson,
  tryMap
} from "../chunk-MM5OEF2H.mjs";
import {
  dataToUUID,
  isValidUUID,
  randomId,
  randomUUID
} from "../chunk-VNNKZIFB.mjs";
import "../chunk-BANDZXMP.mjs";
import "../chunk-MSUB6DGR.mjs";
import "../chunk-IAFBVORQ.mjs";
import "../chunk-FCCOSO5L.mjs";
import {
  executeConditions
} from "../chunk-PL5WNHFZ.mjs";
import {
  actionParametersToJsonSchema,
  convertJsonSchemaToZodSchema,
  jsonSchemaToActionParameters
} from "../chunk-2KQ6HEWZ.mjs";
import "../chunk-GYZIHHE6.mjs";
import "../chunk-P7STFMPO.mjs";
import "../chunk-LGWATNA5.mjs";
import "../chunk-6QGXWNS5.mjs";
import "../chunk-NAFEBKSO.mjs";
export {
  COPILOT_CLOUD_ERROR_NAMES,
  ConfigurationError,
  CopilotKitAgentDiscoveryError,
  CopilotKitApiDiscoveryError,
  CopilotKitError,
  CopilotKitErrorCode,
  CopilotKitLowLevelError,
  CopilotKitMisuseError,
  CopilotKitRemoteEndpointDiscoveryError,
  CopilotKitVersionMismatchError,
  ERROR_CONFIG,
  ERROR_NAMES,
  MissingPublicApiKeyError,
  ResolvedCopilotKitError,
  Severity,
  UpgradeRequiredError,
  actionParametersToJsonSchema,
  convertJsonSchemaToZodSchema,
  dataToUUID,
  executeConditions,
  getPossibleVersionMismatch,
  isMacOS,
  isValidUUID,
  jsonSchemaToActionParameters,
  parseJson,
  randomId,
  randomUUID,
  tryMap
};
