// @copilotkit/sdk-js/dist/langchain.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/langchain.ts
var langchain_exports = {};
__export(langchain_exports, {
  CopilotKitPropertiesAnnotation: () => CopilotKitPropertiesAnnotation,
  CopilotKitStateAnnotation: () => CopilotKitStateAnnotation,
  convertActionToDynamicStructuredTool: () => convertActionToDynamicStructuredTool,
  convertActionsToDynamicStructuredTools: () => convertActionsToDynamicStructuredTools,
  copilotKitCustomizeConfig: () => copilotkitCustomizeConfig,
  copilotKitEmitMessage: () => copilotkitEmitMessage,
  copilotKitEmitState: () => copilotkitEmitState,
  copilotKitEmitToolCall: () => copilotkitEmitToolCall,
  copilotKitExit: () => copilotkitExit
});
module.exports = __toCommonJS(langchain_exports);

// src/langgraph.ts
var import_dispatch = require("@langchain/core/callbacks/dispatch");
var import_shared = require("@copilotkit/shared");
var import_langgraph = require("@langchain/langgraph");
var import_tools = require("@langchain/core/tools");
var import_messages = require("@langchain/core/messages");
var CopilotKitPropertiesAnnotation = import_langgraph.Annotation.Root({
  actions: import_langgraph.Annotation
});
var CopilotKitStateAnnotation = import_langgraph.Annotation.Root({
  copilotkit: import_langgraph.Annotation,
  ...import_langgraph.MessagesAnnotation.spec
});
function copilotkitCustomizeConfig(baseConfig, options) {
  const metadata = (baseConfig == null ? void 0 : baseConfig.metadata) || {};
  if (options == null ? void 0 : options.emitAll) {
    metadata["copilotkit:emit-tool-calls"] = true;
    metadata["copilotkit:emit-messages"] = true;
  } else {
    if ((options == null ? void 0 : options.emitToolCalls) !== void 0) {
      metadata["copilotkit:emit-tool-calls"] = options.emitToolCalls;
    }
    if ((options == null ? void 0 : options.emitMessages) !== void 0) {
      metadata["copilotkit:emit-messages"] = options.emitMessages;
    }
  }
  if (options == null ? void 0 : options.emitIntermediateState) {
    const snakeCaseIntermediateState = options.emitIntermediateState.map((state) => ({
      tool: state.tool,
      tool_argument: state.toolArgument,
      state_key: state.stateKey
    }));
    metadata["copilotkit:emit-intermediate-state"] = snakeCaseIntermediateState;
  }
  baseConfig = baseConfig || {};
  return {
    ...baseConfig,
    metadata
  };
}
__name(copilotkitCustomizeConfig, "copilotkitCustomizeConfig");
async function copilotkitExit(config) {
  await (0, import_dispatch.dispatchCustomEvent)("copilotkit_exit", {}, config);
}
__name(copilotkitExit, "copilotkitExit");
async function copilotkitEmitState(config, state) {
  await (0, import_dispatch.dispatchCustomEvent)("copilotkit_manually_emit_intermediate_state", state, config);
}
__name(copilotkitEmitState, "copilotkitEmitState");
async function copilotkitEmitMessage(config, message) {
  await (0, import_dispatch.dispatchCustomEvent)("copilotkit_manually_emit_message", {
    message,
    message_id: (0, import_shared.randomId)(),
    role: "assistant"
  }, config);
}
__name(copilotkitEmitMessage, "copilotkitEmitMessage");
async function copilotkitEmitToolCall(config, name, args) {
  await (0, import_dispatch.dispatchCustomEvent)("copilotkit_manually_emit_tool_call", {
    name,
    args,
    id: (0, import_shared.randomId)()
  }, config);
}
__name(copilotkitEmitToolCall, "copilotkitEmitToolCall");
function convertActionToDynamicStructuredTool(actionInput) {
  return new import_tools.DynamicStructuredTool({
    name: actionInput.name,
    description: actionInput.description,
    schema: (0, import_shared.convertJsonSchemaToZodSchema)(actionInput.parameters, true),
    func: async () => {
      return "";
    }
  });
}
__name(convertActionToDynamicStructuredTool, "convertActionToDynamicStructuredTool");
function convertActionsToDynamicStructuredTools(actions) {
  return actions.map((action) => convertActionToDynamicStructuredTool(action));
}
__name(convertActionsToDynamicStructuredTools, "convertActionsToDynamicStructuredTools");

// src/langchain.ts
console.warn("Warning: '@copilotkit/sdk-js/langchain' is deprecated and will be removed in a future release. Please use '@copilotkit/sdk-js/langgraph' instead.");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CopilotKitPropertiesAnnotation,
  CopilotKitStateAnnotation,
  convertActionToDynamicStructuredTool,
  convertActionsToDynamicStructuredTools,
  copilotKitCustomizeConfig,
  copilotKitEmitMessage,
  copilotKitEmitState,
  copilotKitEmitToolCall,
  copilotKitExit
});
